File: 백준/Gold/2610. 회의준비/회의준비.js
const input = require("fs").readFileSync("/dev/stdin", "utf-8").trim().split("\n");

const n = Number(input[0]);
const m = Number(input[1]);
const dist = Array.from({ length: n + 1 }, () => Array(n + 1).fill(Infinity));

// 자기 자신까지 거리 0
for (let i = 1; i <= n; i++) {
	dist[i][i] = 0;
}

for (let i = 2; i < 2 + m; i++) {
	const [a, b] = input[i].trim().split(" ").map(Number);
	dist[a][b] = 1;
	dist[b][a] = 1;
}

// 플로이드-워셜
for (let k = 1; k <= n; k++) {
	for (let i = 1; i <= n; i++) {
		for (let j = 1; j <= n; j++) {
			if (dist[i][k] + dist[k][j] < dist[i][j]) {
				dist[i][j] = dist[i][k] + dist[k][j];
			}
		}
	}
}

// 그룹 나누기
const visited = Array(n + 1).fill(false);
const groups = [];

for (let i = 1; i <= n; i++) {
	if (!visited[i]) {
		const group = [];
		const queue = [i];
		visited[i] = true;

		while (queue.length > 0) {
			const cur = queue.shift();
			group.push(cur);

			for (let j = 1; j <= n; j++) {
				if (!visited[j] && dist[cur][j] !== Infinity) {
					visited[j] = true;
					queue.push(j);
				}
			}
		}

		groups.push(group);
	}
}

// 대표자 선정
const representatives = [];

for (const group of groups) {
	let minScore = Infinity;
	let representative = -1;

	for (const person of group) {
		let score = 0;
		for (const other of group) {
			score = Math.max(score, dist[person][other]);
		}

		if (score < minScore) {
			minScore = score;
			representative = person;
		}
	}

	representatives.push(representative);
}

representatives.sort((a, b) => a - b);

// 출력
console.log(representatives.length);
for (const rep of representatives) {
	console.log(rep);
}

File: 백준/Gold/2610. 회의준비/회의준비.py

from collections import deque



n = int(input())
m = int(input())

INF = int(1e9)
dist = [[INF] * (n + 1) for _ in range(n + 1)]

for i in range(1, n + 1):
    dist[i][i] = 0

for _ in range(m):
    a, b = map(int, input().split())
    dist[a][b] = 1
    dist[b][a] = 1

# Floyd-Warshall
for k in range(1, n + 1):
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            if dist[i][k] + dist[k][j] < dist[i][j]:
                dist[i][j] = dist[i][k] + dist[k][j]

# 그룹 찾기 (BFS)
visited = [False] * (n + 1)
groups = []

for i in range(1, n + 1):
    if not visited[i]:
        group = []
        queue = deque([i])
        visited[i] = True

        while queue:
            cur = queue.popleft()
            group.append(cur)

            for j in range(1, n + 1):
                if dist[cur][j] != INF and not visited[j]:
                    visited[j] = True
                    queue.append(j)

        groups.append(group)

# 대표자 선정
representatives = []
for group in groups:
    min_score = INF
    rep = -1

    for person in group:
        score = max(dist[person][other] for other in group)

        if score < min_score:
            min_score = score
            rep = person

    representatives.append(rep)

# 대표자 정렬 및 출력
representatives.sort()

print(len(representatives))
for rep in representatives:
    print(rep)
